url354ce1a2d1_dagaren_gladiator_chess_tgz-pJ8-es_dagaren_gladiator_main_MainJ8/src/es/dagaren/gladiator/representation/BitboardPosition.java

```diff
@@ -0,0 +1,2907 @@
+   {
+      //long before = System.currentTimeMillis();
+      movesList.clear();
+      
+      bbTurnOccupied = bbColourOccupation[turn.index];
+      bbOppositeOccupied = bbColourOccupation[turn.opposite().index];
+      
+      //Se recupera la posición y bitboard del rey
+      bbKing = bbPiecesOccupation[Piece.get(GenericPiece.KING, turn).index];
+      kingSquareIndex = Long.numberOfTrailingZeros(bbKing);
+      
+      if(isInCheck(turn))
+      {
+         //Se generan los movimientos si el bando al que le toca jugar
+         //se encuentra en jaque
+         
+         
+         //Se generan los movimientos del rey a casillas que no están atacadas por las piezas
+         //del oponente.
+         /************** GENERACIÓN DE LOS MOVIMIENTOS DE REY *************/
+         piece = Piece.get(GenericPiece.KING, turn);
+         bbPieceOccupation = bbPiecesOccupation[piece.index];
+         numPieces = Long.bitCount(bbPieceOccupation);
+         for(int i = 0; i < numPieces; i++)
+         {
+            pieceSquareIndex = Long.numberOfTrailingZeros(bbPieceOccupation);
+            
+            //Se obtienen las casillas que ataca la pieza
+            bbAttacked = bbAttackedFromSquare[pieceSquareIndex];
+            bbAttacked &= ~bbTurnOccupied;
+      
+            //Se eliminan de las casillas en las que se mantiene el jaque por alguna pieza. 
+            bbAttackers = bbAttackerToSquare[pieceSquareIndex] & bbOppositeOccupied;
+
+            numAttackers = Long.bitCount(bbAttackers);
+
+            for(int j = 0; j < numAttackers; j++)
+            {
+               attackerSquareIndex = Long.numberOfTrailingZeros(bbAttackers);
+
+               if(pieceInSquare[attackerSquareIndex].genericPiece == GenericPiece.PAWN)
+                 bbAttacked &= ~(bbInclusiveConnection[attackerSquareIndex][kingSquareIndex] & bbAttacked & ~bbSquare[attackerSquareIndex]);
+               else
+                 bbAttacked &= ~(bbFullConnection[attackerSquareIndex][kingSquareIndex] & bbAttacked & ~bbSquare[attackerSquareIndex]);
+               
+               bbAttackers ^= bbSquare[attackerSquareIndex];
+            }
+
+            numAttacked = Long.bitCount(bbAttacked);
+            for(int j = 0; j < numAttacked; j++)
+            {
+               attackedSquareIndex = Long.numberOfTrailingZeros(bbAttacked);
+               
+               //Sólo se añade el movimiento si la casilla destino no está atacada por una pieza del contrario
+               if((bbAttackerToSquare[attackedSquareIndex] & bbColourOccupation[turn.opposite().index]) == 0)
+               {
+                  capture = pieceInSquare[attackedSquareIndex];
+                  mov = new Movement(square[pieceSquareIndex], square[attackedSquareIndex], piece, capture);
+                  
+                  movesList.add(mov);
+               }
+               
+               
+               bbAttacked ^= bbSquare[attackedSquareIndex];
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1763:1778, 2341:2355.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.